### DISCUSSION TITLE
Setting up using NGINX with DuckDNS

### DISCUSSION THREAD

**Discussion by peterhow11** [llamafilm/tesla-http-proxy-addon#14] [General]
Hi,

Thanks very much for working on this solution. I admit that a lot of this is a little over my head, however I am persisting. :-)

I previously used the NGINX Proxy Manger for external access to my Home Assistant and successfully registered for a Tesla Developer account using that subdomain. https://hass.example.com. It was immediately approved. 

I then discovered your addon and reconfigured my system to use NGINX Home Assistant SSL proxy and have successfully setup DuckDNS with certificates for https://hass.example.com and https://tesla.example.com. Since the NGINX Proxy UI only allows you to specify one domain (that I can see), I left this as hass.example.com and tested it with external access to Home Assistant, and it all works fine.

I then applied for a new Tesla developer request based on https://tesla.example.com and https://tesla.example.com/callback as your documentation suggested however this was immediately rejected as "Invalid domain. Domain must be registered with a certificate authority".  I assume that this is because it is trying to access the NGINX server to validate the certificate but does get a response since I have not setup the tesla.example.com domain. 

Not knowing what your addon does underneath the hood (frunk?), I am wondering whether I should simply switch the NGINX proxy to tesla.example.com to allow Tesla to valid the certificate, then switch it back to hass.example.com then start your add-on?  I assume that your addon will add the second domain into the NGINX setup?

BTW. I am located in Australia and have a new M3 (as of 4 weeks ago).

Thanks in advance for your feedback.

**Comment by baylanger** [comment]
example.com should not he used to register at Developer Tesla. This is purely an example.

You must register your own domain. That’s the only way you can have a valid SSL certificate.

Unless I misread your comment, I hope the above helps a little.

**Reply by peterhow11** [reply]
Thanks for pointing that out.  I used my real domain name but just used "example" to hide it in this post!

**Comment by llamafilm** [comment]
Tesla seems to be fairly picky about TLS certificate authorities and also which TLDs they will accept.  Please share the output of `nmap -p 443 --script ssl-cert tesla.llamafilm.com` (with your real domain name) so we can see the certificate.

**Reply by peterhow11** [reply]
Thanks for your suggestion. I tried this and the result was:

Starting Nmap 7.80 ( https://nmap.org ) at 2024-03-04 08:06 AWST
Note: Host seems down. If it is really up, but blocking our ping probes, try -Pn

It was the same with both hass.XXXX and tesla.XXXX

Using the -Pn option was better with this result:

Starting Nmap 7.80 ( https://nmap.org ) at 2024-03-04 08:11 AWST
Nmap scan report for tesla.XXXXXXX.com.au (159.196.XXX.XXX)
Host is up.

PORT    STATE    SERVICE
443/tcp filtered https

Nmap done: 1 IP address (1 host up) scanned in 2.27 seconds

Does that help?

**Reply by llamafilm** [reply]
It sounds like your nginx is not configured properly.  You should have output similar to this, showing the certificate info.  It will be much easier to help if you share your real domain names.  It's publicly reachable so there's no need to hide it.  I see you mentioned `.com.au` in one place, so it's possible that TLD may not allowed by Tesla.

```
% nmap -p 443 --script ssl-cert tesla.llamafilm.com
Starting Nmap 7.94 ( https://nmap.org ) at 2024-03-03 15:38 PST
Nmap scan report for tesla.llamafilm.com (47.158.156.189)
Host is up (0.0044s latency).

PORT    STATE SERVICE
443/tcp open  https
| ssl-cert: Subject: commonName=home.llamafilm.com
| Subject Alternative Name: DNS:home.llamafilm.com, DNS:tesla.llamafilm.com
| Issuer: commonName=R3/organizationName=Let's Encrypt/countryName=US
| Public Key type: rsa
| Public Key bits: 2048
| Signature Algorithm: sha256WithRSAEncryption
| Not valid before: 2024-02-04T22:29:50
| Not valid after:  2024-05-04T22:29:49
| MD5:   5807:9eea:0b3e:7f70:be35:9ce2:7e3a:8bb4
|_SHA-1: b187:84cf:234a:36c3:8f19:9e41:551d:53ca:2b1a:bc61
```

**Reply by peterhow11** [reply]
Good point. It is tesla.howletts.com.au. Did you see my post below regarding the Tesla developer application working after I changed NGINX to use this domain?

**Reply by baylanger** [reply]
[edit] never mind you say it works now!!

There are sites to analyze SSL cert, you might want to compare your site vs llamafilm and see if there’s any difference. I did it for your site but just like earlier I’m still away on my iPhone. Aside missing certs in the chain and the signed vendor, it looks like all is good. llamafilm had same output unless I missed something on my small screen.

Perhaps like llamafilm said, maybe Tesla doesn’t accept .au yet?

https://www.sslchecker.com/sslchecker

**Reply by llamafilm** [reply]
Okay, your TLS cert looks fine to me.  I see you're using Lets Encrypt which is known to work.
If hass.howletts.com.au works, then that proves the TLD is okay.
Maybe it does need an actual service listening to verify.  It's possible they may have changed the validation process after I set mine up.
Try installing the addon and running through the config, and once you are able to access https://tesla.howletts.com.au/.well-known/appspecific/com.tesla.3p.public-key.pem then try the registration again.

**Reply by peterhow11** [reply]
Thanks. I have progressed through most stages except these last steps:

1. Test Public Key Endpoint - Shows dialog saying "This website is trying to download a configuration profile. Do you want to allow this?" Clicking Allow shows dialog "Profile Error - Invalid profile".

2. Shut down Flash Server -   This shows  Error 502: Bad gateway on the page and a dialog message "The add-on seems not to be ready, it might be still starting. Do you want to try again?". Retrying is a loop to the same.

3. Configure the [Tesla integration](https://github.com/alandtse/tesla) to use this proxy. I am sorry, I can't find any instructions on how to do this. Please point me in the correct direction.

If I wish to retry the whole process, do I need to do more than uninstall and reinstall the Addon?

FYI. Here is the log file after re-starting the Tesla HTTP Proxy:

===============================================

s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[16:50:44] INFO: Found existing keypair
[16:50:44] NOTICE: Starting temporary Python app for authentication flow
[16:50:44] main:INFO: *** Generating Partner Authentication Token ***
[16:50:45] main:INFO: *** Registering Tesla account ***
[16:50:47] main:INFO: {"response":{"account_id":"48acb8b0-8555-4c86-9eb3-95e765ba1fe9","domain":"tesla.howletts.com.au","name":"Home Assistant V2","description":"Allow integration with Home Assistant","csr":null,"client_id":"8c2bd48b83d8-457a-b535-5d3ce0e3c2ec","ca":null,"created_at":"2024-03-04T01:35:06.558Z","updated_at":"2024-03-04T01:35:06.558Z","enterprise_tier":"free","issuer":null,"csr_updated_at":null,"public_key":"0486cf03b912d7e6e4db3f0e513a6871392df1359d32374c800beb0a378cf2522e74d232f4f2660daff48c88e3532103eccad69d25eb11dbe83481e5d3453a6b80"}}
[16:50:47] main:INFO: *** Starting Flask server... ***
[16:50:47] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.8:8099
[16:50:47] werkzeug:INFO: Press CTRL+C to quit
[16:51:33] werkzeug:INFO: 172.30.32.2 - - [04/Mar/2024 16:51:33] "GET / HTTP/1.1" 200 -
[16:53:58] run:WARNING: Info to enter into Tesla Custom component:
         Refresh token  : xxxxxTOKENxxxxx
         Proxy URL      : https://c03d64a7-tesla-http-proxy
         SSL certificate: /share/tesla/selfsigned.pem
         Client ID      : xxxxxCLIENTIDxxxxxx

[16:53:58] werkzeug:INFO: 172.30.33.7 - - [04/Mar/2024 16:53:58] "GET /callback?locale=en-US&code=NA_072953e705f3306b83c773bc8806723b2024fd5e0544590f3d925e52b9c5&state=d3839e2fae404695a4de246cdabd3809&issuer=https://auth.tesla.com/oauth2/v3 HTTP/1.0" 200 -

[16:54:07] INFO: Starting Tesla HTTP Proxy

Do not listen on a network interface without adding client authentication. Unauthorized clients may
be used to create excessive traffic from your IP address to Tesla's servers, which Tesla may respond
to by rate limiting or blocking your connections.
2024-03-04T16:54:07+08:00 [debug] Creating proxy
2024-03-04T16:54:07+08:00 [info ] Listening on 0.0.0.0:443

===============================================

Thanks again for your time.

**Reply by peterhow11** [reply]
I just tried the "Test Public Key Endpoint" option on my Mac, and it returned a valid-looking key (as per your example). I was previously doing this on iOS and getting the Invalid profile error.  So it seems this part is working.

**Reply by llamafilm** [reply]
This all sounds correct! The Tesla addon with support for this hasn’t been released yet, it’s currently an open PR, I’m almost done.
That’s good feedback about the public key on iOS, I’ll see if I can fix that with a mime type.

**Reply by peterhow11** [reply]
Thanks for clarifying that. It's also probably worth clarifying in the documentation that, during the Tesla developer application process, Telsa now seem to require validation on the SSL status for the nominated domain endpoints. So you need to have this setup in NGINX Home Assistant SSL Proxy prior to completing the application then swap back to the original domain (used for remote access to HA) before starting the Addon (which then adds your specified Telsa domain as a second domain to NGINX).

**Comment by peterhow11** [comment]
I changed the domain name in NGINX to tesla.XXXXX.com and retried the Tesla developer application and it worked!  So, should I change it back to hass.XXXXX.com before starting your Add-on?

**Comment by baylanger** [comment]
Leave it has tesla…. There was a typo I reported earlier m, it’s fixed in the latest Pr.

**Reply by baylanger** [reply]
@peterhow11 just to make sure, you are referring to this?

https://github.com/llamafilm/tesla-http-proxy-addon/commit/173421fa4d930897acfcc8bf96b7e10fb7be26ba#diff-a7c33cf0e04fcea0683d73197a3a058ddcecb6a4c3184318eed72042398a8b78

**Comment by paky79** [comment]
Hello, I am having the same issue with a configuration duckdns+nginx.
My main DNS has this SSL:

PORT    STATE SERVICE
443/tcp open  https
| ssl-cert: Subject: commonName=pakyhome.duckdns.org
| Subject Alternative Name: DNS:pakyhome.duckdns.org
| Issuer: commonName=R3/organizationName=Let's Encrypt/countryName=US
| Public Key type: ec
| Public Key bits: 384
| Signature Algorithm: sha256WithRSAEncryption
| Not valid before: 2024-02-11T18:05:47
| Not valid after:  2024-05-11T18:05:46
| MD5:   e70f 8a13 7013 3040 5b84 93cb e64c 02e7
|_SHA-1: 099e cf39 2b71 2cf0 b936 6cb8 bf6f 3480 1d4f f547

I try to start the add-on putting in the configuration only the second domain teslamypaky.duckdns.org. I check the certificate and I saw it is pointing to the same main certificates for the main DNS:

PORT    STATE SERVICE
443/tcp open  https
| ssl-cert: Subject: commonName=pakyhome.duckdns.org
| Subject Alternative Name: DNS:pakyhome.duckdns.org
| Issuer: commonName=R3/organizationName=Let's Encrypt/countryName=US
| Public Key type: ec
| Public Key bits: 384
| Signature Algorithm: sha256WithRSAEncryption
| Not valid before: 2024-02-11T18:05:47
| Not valid after:  2024-05-11T18:05:46
| MD5:   e70f 8a13 7013 3040 5b84 93cb e64c 02e7
|_SHA-1: 099e cf39 2b71 2cf0 b936 6cb8 bf6f 3480 1d4f f547

This is what I received by the add-on;
[18:55:26] INFO: Found existing keypair
[18:55:26] NOTICE: Starting temporary Python app for authentication flow
[18:55:27] main:INFO: *** Generating Partner Authentication Token ***
[18:55:27] main:ERROR: HTTP 400: Bad Request
Traceback (most recent call last):
  File "/app/run.py", line 46, in <module>
    tesla_api_token = req.json()['access_token']
                      ~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: 'access_token'

If I try to activate the Tesla developer account anyway I receive the error: invalid domain.

I tried also putting the second DNS in NGINX but nothing works. Any idea?

**Reply by baylanger** [reply]
The certificate must match the hostname teslamypaky.duckdns.org.

You can use this site to test your cert [here](https://www.ssllabs.com/ssltest/analyze.html?d=teslamypaky.duckdns.org)

**Reply by paky79** [reply]
This is because everything is pointing to the only certificates I have for the main DNS. Should I release new certificates for the sub-domain? How will be managed by NGINX?

**Reply by baylanger** [reply]
I don’t use DuckDNS but you should be able to request a multi domain certificate (SAN) under the same cert. This is what @peterhow11 has for his SSL cert. 

[edit] if you use letsencrypt to get your cert you can probably adjust the config to fetch multiple names, look around I’m sure someone wrote docs for this.

**Reply by llamafilm** [reply]
@paky79 this sounds like a different issue because you’re getting a Python exception. Please open a separate issue for this and describe detailed steps you took.

**Reply by paky79** [reply]
It seems that at the end I successfully got access as developer! This was fixed with multi-domain certificates! 
I am trying to test the request:
curl --cacert /share/tesla/selfsigned.pem \
    --header "Authorization: Bearer $TESLA_AUTH_TOKEN" \
    "https://addon-tesla-http-proxy/api/1/vehicles"
But I received the error: URL rejected. Malformed input to a URL function

Any suggestion for that?

**Reply by baylanger** [reply]
Just to make sure, you did set the variable TESLA_AUTH_TOLEN to your token? If so I’m not sure what the issue could be.

**Reply by paky79** [reply]
Should I set to refresh token?

**Reply by baylanger** [reply]
I’m not aware of other token, it's likely the refresh token but I’m not positive. If it’s the refresh token, I have a PR to make some changes to the docs and would propose we update the var name. Maybe wait for @llamafilm to confirm.

**Reply by llamafilm** [reply]
May I suggest keeping topics focused with a separate issue/discussion for each topic? It will help make these discussions more useful for future readers.

**Reply by paky79** [reply]
Sorry for that but I think that also this can help to complete the full configuration. If you prefer I will open other discussion/issues. Than you

**Comment by baylanger** [comment]
@peterhow11 you mind editing the title of this dicussion for something like ~ Nginx DuckDNS but... I'm not sure there's a big difference between DuckDNS and privately owned domain, maybe that's not a good title either :/

**Reply by peterhow11** [reply]
done

**Comment by tobik176** [comment]
Hello, I am also struggling with nginx, duckdns and tesla HTTP proxy. It seems like there is a problem with the SSL certificate but I don't know how to fix it :( can anyone help? thanks in advance!

Also I noticed nginx already uses port 443, so for tesla HTTP proxy I had to use another port.


```
[18:21:02] webui:INFO: Starting Flask server for Web UI...
[18:21:02] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.8:8099
[18:21:02] werkzeug:INFO: Press CTRL+C to quit
[18:21:03] INFO: Setting up GnuPG and password-store
gpg: keybox '/data/gnugpg/pubring.kbx' created
gpg: /data/gnugpg/trustdb.gpg: trustdb created
gpg: directory '/data/gnugpg/openpgp-revocs.d' created
gpg: revocation certificate stored as '/data/gnugpg/openpgp-revocs.d/8B5002733A82980E0C3C8E2F07F567B708A82DE1.rev'
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: next trustdb check due at 2027-06-03
/data/gnugpg/pubring.kbx
------------------------
pub   ed25519 2024-06-03 [SC] [expires: 2027-06-03]
      8B5002733A82980E0C3C8E2F07F567B708A82DE1
uid           [ultimate] myself
sub   cv25519 2024-06-03 [E]

created directory: '/data/password-store/'
Password store initialized for myself
[18:21:03] INFO: Current Nginx configuration
{"domain":"haknabe.duckdns.org","hsts":"max-age=31536000; includeSubDomains","certfile":"fullchain.pem","keyfile":"privkey.pem","cloudflare":false,"customize":{"active":true,"default":"nginx_proxy_default*.conf","servers":"nginx_proxy/*.conf"}}
[18:21:04] INFO: Adding custom config to /share/nginx_proxy/nginx_tesla.conf
    server {
        server_name teslaknabe.duckdns.org;

        ssl_certificate /ssl/fullchain.pem;
        ssl_certificate_key /ssl/privkey.pem;

        # dhparams file
        ssl_dhparam /data/dhparams.pem;

        listen 443 ssl http2;
        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;

        location / {
            return 404;
        }

        # static public key for Tesla
        location /.well-known/appspecific/com.tesla.3p.public-key.pem {
            root /share/tesla;
            try_files /com.tesla.3p.public-key.pem =404;
        }

        location = /favicon.ico {
            log_not_found off;
        }

        location = /robots.txt {
            log_not_found off;
        }
    }
[18:21:04] INFO: Generating self-signed SSL certificate
-----
[18:21:04] INFO: Generating keypair
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5FWzrKD6d9zdewZJjlDxWkUqVZa8
eg8ad13EvpiRno9OqTCn2PHhnbnGFzCZynomX8EDn+g/BXI2Iii7cC5G6Q==
-----END PUBLIC KEY-----
[18:21:04] INFO: Testing public key...
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 1 seconds. 14 retries left.
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 2 seconds. 13 retries left.
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 4 seconds. 12 retries left.
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 8 seconds. 11 retries left.
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 16 seconds. 10 retries left.
curl: (6) Could not resolve host: teslaknabe.duckdns.org
Warning: Problem : timeout. Will retry in 32 seconds. 9 retries left.
```

```
[18:42:13] webui:INFO: Starting Flask server for Web UI...
[18:42:13] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.8:8099
[18:42:13] werkzeug:INFO: Press CTRL+C to quit
[18:42:14] INFO: Found existing keypair
[18:42:14] INFO: Testing public key...
curl: (60) SSL certificate problem: self-signed certificate in certificate chain
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Warning: Problem (retrying all errors). Will retry in 1 seconds. 14 retries 
Warning: left.
curl: (60) SSL certificate problem: self-signed certificate in certificate chain
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Warning: Problem (retrying all errors). Will retry in 2 seconds. 13 retries 
Warning: left.
```

**Comment by baylanger** [comment]
Externally your cert looks good per the results found here https://www.ssllabs.com/ssltest/analyze.html?d=teslaknabe.duckdns.org

Do you use the nginx addon? How is port 443 exposed to the world; Port forwarding from router to your HA server or something else?

curl complains about a self-signed cert… do you have any services that exposes such a cert?

**Reply by tobik176** [reply]
Hi, thanks for your message.

Yes, I use the NGINX Home Assistant SSL proxy addon.

I use port forwarding from my router and I use port 443 for nginx and port 4430 for tesla HTTP proxy addon. I am not sure if that's correct...

I use DuckDNS addon which brings Let's Encrypt SSL with it.

**Reply by baylanger** [reply]
The ports look just right. From your PC connected to your local network (WiFi I suppose), using the URL 1) can you download the key? (Look into the instruction how to download the key) 2) is the cert valid? Unless you’ve put the proper rules in place in your router, a local dns most likely needs an entry for “teslaknabe” that points to the local IP of HA.

I’m unclear if the above will help you much but let’s see.
